File: src/main/java/io/github/bot/terminal/application/domains/monitoring/rest/MonitoringTasksRestController.java
package io.github.bot.terminal.application.domains.monitoring.rest;


import io.github.bot.terminal.application.domains.common.conditions.dto.ConditionTypeDTO;
import io.github.bot.terminal.application.domains.common.cron.dto.CronFrequencyDTO;
import io.github.bot.terminal.application.domains.monitoring.rest.dto.MonitoringTaskDTO;
import io.github.bot.terminal.application.domains.monitoring.rest.dto.MonitoringTaskTypeDTO;
import io.github.bot.terminal.application.domains.monitoring.rest.requests.MonitoringTaskRequest;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;
import java.util.Collection;
import java.util.List;

@RestController
@Slf4j
@RequestMapping("/api/monitoring/tasks")
@RequiredArgsConstructor
public class MonitoringTasksRestController {

    private final MonitoringTasksRestService monitoringTasksRestService;

    @PostMapping
    @ResponseStatus(code = HttpStatus.CREATED)
    public MonitoringTaskDTO add(@RequestBody @Valid MonitoringTaskRequest request) {
        return monitoringTasksRestService.add(request);
    }

    @GetMapping
    @ResponseStatus(code = HttpStatus.OK)
    public List<MonitoringTaskDTO> list() {
        return monitoringTasksRestService.list();
    }

    @GetMapping("/{id}")
    @ResponseStatus(code = HttpStatus.OK)
    public MonitoringTaskDTO get(@PathVariable String id) {
        return monitoringTasksRestService.get(id);
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(code = HttpStatus.NO_CONTENT)
    public void delete(@PathVariable String id) {
        monitoringTasksRestService.delete(id);
    }

    @PutMapping("/{id}")
    @ResponseStatus(code = HttpStatus.OK)
    public MonitoringTaskDTO edit(@PathVariable String id,
                                  @RequestBody @Valid MonitoringTaskRequest request) {
        return monitoringTasksRestService.edit(id, request);
    }

    @GetMapping("/cron-frequencies")
    @ResponseStatus(code = HttpStatus.OK)
    public List<CronFrequencyDTO> cronFrequencies() {
        return monitoringTasksRestService.cronFrequencies();
    }

    @GetMapping("/conditions")
    @ResponseStatus(code = HttpStatus.OK)
    public List<ConditionTypeDTO> conditionTypes() {
        return monitoringTasksRestService.conditionTypes();
    }

    @GetMapping("/types")
    @ResponseStatus(code = HttpStatus.OK)
    public Collection<MonitoringTaskTypeDTO> types() {
        return monitoringTasksRestService.types();
    }

}

File: src/main/kotlin/io/github/bot/terminal/application/domains/monitoring/rest/MonitoringTasksRestController.kt
package io.github.bot.terminal.application.domains.monitoring.rest

import io.github.bot.terminal.application.domains.common.conditions.dto.ConditionTypeDTO
import io.github.bot.terminal.application.domains.common.cron.dto.CronFrequencyDTO
import io.github.bot.terminal.application.domains.monitoring.rest.dto.MonitoringTaskDTO
import io.github.bot.terminal.application.domains.monitoring.rest.dto.MonitoringTaskTypeDTO
import io.github.bot.terminal.application.domains.monitoring.rest.requests.MonitoringTaskRequest
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*
import javax.validation.Valid

@RestController
@RequestMapping("/api/monitoring/tasks")
class MonitoringTasksRestController(
        private val monitoringTasksRestService: MonitoringTasksRestService
) {
    @PostMapping
    @ResponseStatus(code = HttpStatus.CREATED)
    fun add(@RequestBody request: @Valid MonitoringTaskRequest): MonitoringTaskDTO {
        return monitoringTasksRestService.add(request)
    }

    @GetMapping
    @ResponseStatus(code = HttpStatus.OK)
    fun list(): List<MonitoringTaskDTO> {
        return monitoringTasksRestService.list()
    }

    @GetMapping("/{id}")
    @ResponseStatus(code = HttpStatus.OK)
    operator fun get(@PathVariable id: String): MonitoringTaskDTO {
        return monitoringTasksRestService[id]
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(code = HttpStatus.NO_CONTENT)
    fun delete(@PathVariable id: String) {
        monitoringTasksRestService.delete(id)
    }

    @PutMapping("/{id}")
    @ResponseStatus(code = HttpStatus.OK)
    fun edit(@PathVariable id: String,
             @RequestBody request: @Valid MonitoringTaskRequest): MonitoringTaskDTO {
        return monitoringTasksRestService.edit(id, request)
    }

    @GetMapping("/cron-frequencies")
    @ResponseStatus(code = HttpStatus.OK)
    fun cronFrequencies(): List<CronFrequencyDTO> {
        return monitoringTasksRestService.cronFrequencies()
    }

    @GetMapping("/conditions")
    @ResponseStatus(code = HttpStatus.OK)
    fun conditionTypes(): List<ConditionTypeDTO> {
        return monitoringTasksRestService.conditionTypes()
    }

    @GetMapping("/types")
    @ResponseStatus(code = HttpStatus.OK)
    fun types(): Collection<MonitoringTaskTypeDTO> {
        return monitoringTasksRestService.types()
    }
}